### PR TITLE
Move imp validation into ortb package

### PR DESCRIPTION
**Pull request by bsardo**
This PR begins the process of moving all of the validation logic out of the `endpoints/openrtb2` package and into the `ortb` package. A request validator object is instantiated in the router and then injected into the auction, video and amp endpoints where it is available for use to validate imps initially.

The driver of this change is that we want to be able to leverage the existing imp validation logic after we have merged imp first party data. In this case, validation needs to be re-run on impressions only. 

As part of this process, additional unit tests were added for imp banner, video, audio and pmp, though native and ext are not complete. We do have code coverage for these though through our E2E endpoint tests.

### CODE REVIEW

File: ortb/request_validator.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	if err := rv.validateStoredBidResponses(prebid, storedBidResp, imp.ID); err != nil {
+		return []error{err}
+	}
+
+	errL := []error{}
+
+	for bidder, ext := range prebid.Bidder {
+		coreBidder, _ := openrtb_ext.NormalizeBidderName(bidder)
```
Comment (SyntaxNode) [inline ortb/request_validator.go:110]: Can we pull `NormalizeBidderName` to a dependency on the object?

File: endpoints/openrtb2/amp_auction.go [hunk 2]
```diff
@@ -60,6 +60,7 @@ func NewAmpEndpoint(
 	uuidGenerator uuidutil.UUIDGenerator,
 	ex exchange.Exchange,
 	validator openrtb_ext.BidderParamValidator,
```
Comment (SyntaxNode) [inline endpoints/openrtb2/amp_auction.go:62]: Do we still need `paramsValidator` with it now incorporated into the `requestValidator`?

File: ortb/request_validator.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+		}
+	}
+	return nil
+}
+
+func generateStoredBidResponseValidationError(impID string) error {
+	return fmt.Errorf("request validation failed. Stored bid responses are specified for imp %s. Bidders specified in imp.ext should match with bidders specified in imp.ext.prebid.storedbidresponse", impID)
+}
+
+// TODO: move this outside of this package
```
Comment (bsardo) [inline ortb/request_validator.go:174]: Where should this function live? It was originally in `endpoints/openrtb2/auction.go` and was being used for both imp validation and in the function `mergeBidderParamsImpExt` which were both in the same package. For now I just copied the function here but this seems like an awkward place for it. Perhaps it belongs in some package other than `ortb` and `endpoints/openrtb2`?

Comment (bsardo) [reply ortb/request_validator.go:174]: Discussed and can move into `bidders.go#IsPotentialBidder` below `IsBidderNameReserved`.

File: ortb/request_validator_banner.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		}
+	}
+
+	return nil
+}
+
+func validateFormat(format *openrtb2.Format, impIndex, formatIndex int) error {
+	if format == nil {
+		return nil
+	}
```
Comment (bsardo) [inline ortb/request_validator_banner.go:60]: The functions in this file were copied and this nil check was added.

File: ortb/request_validator_video.go
Comment (bsardo) [inline ortb/request_validator_video.go:1]: The functions in this file were copied.

File: ortb/request_validator_pmp.go
Comment (bsardo) [inline ortb/request_validator_pmp.go:1]: The functions in this file were copied.

File: ortb/request_validator_audio.go
Comment (bsardo) [inline ortb/request_validator_audio.go:1]: The functions in this file were copied.

File: ortb/request_validator_native.go
Comment (bsardo) [inline ortb/request_validator_native.go:1]: The functions in this file were copied.

File: ortb/request_validator.go
Comment (bsardo) [inline ortb/request_validator.go:1]: The functions in this file were copied with the ones that were endpoint deps receiver methods becoming receiver methods on the newly created `RequestValidator` type.

File: ortb/request_validator_native_test.go
Comment (bsardo) [inline ortb/request_validator_native_test.go:1]: These tests were copied.

File: openrtb_ext/bidders.go [hunk 5]
```diff
@@ -555,6 +577,8 @@ var bidderNameLookup = func() map[string]BidderName {
 	return lookup
 }()
 
+type NameNormalizer func(name string) (BidderName, bool)
```
Comment (SyntaxNode) [inline openrtb_ext/bidders.go:580]: Nitpick: I think too much context is lost with the dropping of "bidder". Consider naming "BidderNameNormalizer" or "BidderNameFormatter".

File: ortb/request_validator.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+	"github.com/prebid/prebid-server/v2/stored_responses"
+)
+
+type RequestValidator struct {
```
Comment (SyntaxNode) [inline ortb/request_validator.go:12]: Consider changing RequestValidator to an interface with only the ValiateImp method for now. Then rename this "standardRequestValidator" to be responded by `NewRequestValidator`.

File: endpoints/openrtb2/amp_auction.go [hunk 7]
```diff
@@ -73,7 +73,7 @@ func NewAmpEndpoint(
 	tmaxAdjustments *exchange.TmaxAdjustmentsPreprocessed,
 ) (httprouter.Handle, error) {
 
-	if ex == nil || validator == nil || requestsById == nil || accounts == nil || cfg == nil || metricsEngine == nil {
```
Comment (hhhjort) [inline endpoints/openrtb2/amp_auction.go:76]: Why do we no longer need this nill check?

Comment (bsardo) [reply endpoints/openrtb2/amp_auction.go:76]: Ah that was removed when I had the validator as a struct. I re-added the nil checks now that it is an interface.

**Review by VeronikaSolovei9 — APPROVED** [review]
Looks good to me

**Review by hhhjort — APPROVED** [review]
LGTM

**Pull request merged**